src/OpenTelemetry.Api/Context/Propagation/TraceContextPropagator.cs

```diff
@@ -220,91 +220,180 @@ internal static bool TryExtractTraceparent(string traceparent, out ActivityTrace
         return true;
     }
 
-    internal static bool TryExtractTracestate(string[] tracestateCollection, out string tracestateResult)
+    internal static bool TryExtractTracestate(IEnumerable<string> tracestateCollection, out string tracestateResult)
     {
         tracestateResult = string.Empty;
 
-        if (tracestateCollection != null)
+        char[] array = null;
+        Span<char> buffer = stackalloc char[256];
+        Span<char> keyLookupBuffer = stackalloc char[96]; // 3 x 32 keys
```

Comment:
Love what this PR is doing but I have some concern here with allocating 704 bytes on the stack `((256 + 96) * 2)`.

System.Text.Json for example will at max `stackalloc` 256 bytes (128 `char`s): https://github.com/dotnet/runtime/blob/a86987ccab917433d065fe5dc8870fc261f79d14/src/libraries/System.Text.Json/Common/JsonConstants.cs#L12-L13

Not sure why _that_ number, but I'm sure a lot of thought went into it 😄 /cc @stephentoub

I think this pattern for "`stackalloc` with fallback to rented array" is solid but the version here seems a bit off from what I have seen. IMO it more commonly looks like this:

```csharp
char[]? rentedBuffer = null;
Span<char> destination = length <= Constants.StackallocCharThreshold ?
   stackalloc char[Constants.StackallocCharThreshold] :
   (rentedBuffer = ArrayPool<char>.Shared.Rent(length));

try
{
   ...
}
finally
{
   if (rentedBuffer != null)
      ArrayPool<char>.Shared.Return(rentedBuffer );
}
```

Would something like that work/help simplify things here?

Reply:
> Not sure why that number, but I'm sure a lot of thought went into it 😄

It's a bit squishy. We almost never go above 1K. We typically use 256 or 512 bytes, but it varies case-to-case based on knowledge of that particular location and how likely longer buffers are expected to be needed.

Reply:
@CodeBlanch I think I'd seen somewhere that 1K was a "reasonable" max value. I went for 256 chars as, realistically, that should cover most trace state scenarios. We could drop to 128 (256B) instead, though, and still cover most shorter trace state strings. Combined with the 96 chars (192B)  for the duplicate lookup, that might be more reasonable.

I can switch to the try/finally here. The' Return' method initially had some extra logic, but I refactored that before the PR. I left it in that form because I slightly preferred avoiding the indentation introduced with the extra blocks.

Reply:
In the [pattern](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5749#discussion_r1678208947) the `try-finally` can be ommited. In case of exceptions the buffer (if rented) will just be dropped instead of returned to the pool, but that isn't a problem to the pool.

In first incarnations of this pattern try-finally got used, but later on that pattern evolved.

Reply:
I don't care so much about the `try-finally` we could drop that. Just want to make sure the `stackalloc` is good.

We can't really control _where_ `Extract` runs so it is probably a good idea to be conservative. For incoming request (think AspNetCore instrumentation) it will probably be ~early. But something like processing messages from a queue, who knows! What I would _really_ like to see is us avoid the `stackalloc` if we know there is a lot of tracestate but I guess hard to do because there could be multiple headers needing to be processed?

Reply:
@CodeBlanch We could just be safe and go with `ArrayPool`. I preferred `stackalloc` since we might reasonably expect the state to be small, and we can avoid a small amount of overhead that we incur by renting.

Reply:
@gfoidl I think we need the `finally` here, though, as there are multiple branches where this code could return from this method on the non-exception path, and we need to ensure that the array is returned in those cases, too.